### DISCUSSION TITLE
Is it possible to go through the pdf page by page once rendered and before saving it?

### DISCUSSION THREAD

**Discussion by rbellette** [dompdf/dompdf#3140] [General]
I would like to be able to go through the rendered PDF to do, for example, on each page that includes a different image for a hypothetical header via the following code:

```
        $imageURL = '/var/www/html/wp-content/themes/my-theme/images/pdf_header.png';

	$imgWidth = 500; 
	$imgHeight = 250;

	// Set image opacity 
	$canvas->set_opacity(.5); 

	// Specify horizontal and vertical position 
	$x = (($w-$imgWidth)/2); 
	$y = (($h-$imgHeight)/3); 

	$canvas->image($imageURL, 0, 0, $imgWidth, $imgHeight, $resolution = "normal"); 
```

The function `$canvas = $dompdf->getCanvas();` It always runs on the last page of the PDF, but I don't know how to tell it to do it in a loop for all pages of the rendered PDF.

Is it possible to do this?

Thanks!!!!

**Comment by bsweeney** [comment]
The `image` method only runs once. If you want to run something on every page one way to do that is to wrap it in the page_script method:

```
$canvas->page_script(function ($pageNumber, $pageCount, $canvas, $fontMetrics) {
    $w = $canvas->get_width();
    $h = $canvas->get_height();

    $imgWidth = 500;
    $imgHeight = 250;

    // Set image opacity
    $canvas->set_opacity(.5);

    // Specify horizontal and vertical position
    $x = (($w-$imgWidth)/2);
    $y = (($h-$imgHeight)/3);

    $canvas->image($imageURL, 0, 0, $imgWidth, $imgHeight, $resolution = "normal");
    $canvas->set_opacity(1);
});
```

Though you're not doing anything the requires scripting here so you might consider just using a fixed-position element. For example:
```
<div style="position: fixed; height: 150px; width: 150px; left: 0px; top: 0px; opacity: .5"><img src="..." style="width: 150px; height: 150px;">
```

**Reply by bsweeney** [reply]
FYI, almost identical question was asked in #3137. You could probably have found the info you needed with a search.

**Comment by rbellette** [comment]
Thanks a lot! 
Everything has worked perfectly. Doing tests I managed to get it to work the day before.

**Comment by rbellette** [comment]
How can you pass a variable from outside the function to page-script?

`
$id = 123;
$canvas = $dompdf->getCanvas();
$canvas->page_script(function ($pageNumber, $pageCount, $canvas, $fontMetrics) {
      $txt = 'Lorem...' . $id;
...`

**Comment by rbellette** [comment]
It's ok. I have found the solution.

`$canvas->page_script(function ($pageNumber, $pageCount, $canvas, $fontMetrics) use ($myVar) {...`
